pkg/kubelet/userns_manager.go

```diff
@@ -522,3 +522,82 @@ func (m *usernsManager) CleanupOrphanedPodUsernsAllocations(pods []*v1.Pod, runn
 
 	return nil
 }
+
+// getHostIDsForPod if the pod uses user namespaces, takes the uid and gid
+// inside the container and returns the host UID and GID those are mapped to on
+// the host. If containerUID is nil, then it returns the host UID for UID 0
+// inside the container. If containerGID is nil, then it returns nil.
+// If the pod is not using user namespaces, as there is no mapping needed, the
+// same containerUID and containerGID params are returned.
+func (m *usernsManager) getHostIDsForPod(pod *v1.Pod, containerUID, containerGID *int64) (hostUID, hostGID *int64, err error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return containerUID, containerGID, nil
+	}
+
+	if pod == nil || pod.Spec.HostUsers == nil || *pod.Spec.HostUsers == true {
+		return containerUID, containerGID, nil
+	}
+
+	mapping, err := m.GetOrCreateUserNamespaceMappings(pod)
+	if err != nil {
+		err = fmt.Errorf("Error getting pod user namespace mapping: %w", err)
+		return
+	}
+	uids := mapping.Uids
+	gids := mapping.Gids
+
+	uid, err := hostIDFromMapping(uids, containerUID)
+	if err != nil {
+		err = fmt.Errorf("Error getting host UID: %w", err)
+		return
+	}
+
+	/* XXX: If there is not GID set, then do not force any.
+	 * Returning a non-nil hostGID will force the fsGroup on the volume,
+	 * which forces group permissions even if the mode asked is 0600. That
+	 * behaviour makes sense when fsGroup is requested by the user, but not
+	 * in this case that it isn't (it is nil).
+	 * This allows, for example, to have secrets with permission 0600 (as
+	 * ssh and other apps enforce on some files) when userns is enabled too.
+	 */
+	if containerGID == nil {
+		return &uid, nil, nil
+	}
+
+	gid, err := hostIDFromMapping(gids, containerGID)
+	if err != nil {
+		err = fmt.Errorf("Error getting host GID: %w", err)
+		return
+	}
+
+	return &uid, &gid, nil
+}
+
+func hostIDFromMapping(mapping []*runtimeapi.IDMapping, containerId *int64) (int64, error) {
+	if mapping == nil {
```

Comment:
`if len(mapping) == 0` ? (to support empty but non-nil case)

Reply:
sure, I'll add that